File: ts/packages/ts/src/ConfigSet.ts
import _ from 'lodash';
import { type IConfigSet } from './types';
import { ERR, NAME } from './utils';

export class ConfigSet implements IConfigSet {
  static SEPARATOR = '/' as const;
  static ROOT = '' as const;
  static ROOT_LABEL = '/' as const;

  public readonly hierarchy: string[] = [];

  constructor(public readonly path: string = ConfigSet.ROOT) {
    const scopeLocation = [`ConfigSet`, `constructor`];

    if (this.path.startsWith(ConfigSet.ROOT_LABEL)) {
      this.path = this.path.slice(1);
    }

    if (this.path.endsWith(ConfigSet.SEPARATOR)) {
      throw new Error(
        ERR(`invalid path "${path}"`, {
          location: scopeLocation,
          suggestion: `path mustn't end with ${ConfigSet.SEPARATOR} character`,
        }),
      );
    }

    if (this.path === ConfigSet.ROOT) {
      this.hierarchy = [ConfigSet.ROOT];
      return;
    }

    this.hierarchy = this.path.split(ConfigSet.SEPARATOR).map((cur, idx, sets) => {
      if (!NAME(cur)) {
        throw new Error(
          ERR(`invalid path "${path}"`, {
            location: scopeLocation,
            suggestion: `path nodes mustn't contain reserved words "${cur}"`,
          }),
        );
      }
      return _.take(sets, idx + 1).join(ConfigSet.SEPARATOR);
    });
    this.hierarchy.unshift(ConfigSet.ROOT);
  }
}

File: py/src/configu/config_set.py
from typing import List

from .config_schema import ConfigSchema
from .types.generated import ConfigSet as IConfigSet
from .utils import pretty_error


class ConfigSet(IConfigSet):
    SEPARATOR: str = "/"
    ROOT: str = ""
    ROOT_LABEL: str = "/"

    def __init__(self, path: str = ROOT):
        self.hierarchy: List[str] = []

        if path.startswith(ConfigSet.ROOT_LABEL):
            path = path[1:]

        if path.endswith(ConfigSet.SEPARATOR):
            err_metadata = {
                "location": ["ConfigSet", "constructor"],
                "suggestion": f"path mustn't end with {ConfigSet.SEPARATOR} character",
            }
            raise ValueError(
                pretty_error(f'invalid path "{path}"', err_metadata)
            )

        if path == ConfigSet.ROOT:
            self.hierarchy = [ConfigSet.ROOT]
            return

        sets = path.split(ConfigSet.SEPARATOR)

        self.hierarchy = [
            ConfigSet.SEPARATOR.join(sets[:i])
            for i, cur in enumerate(sets, 1)
            if ConfigSchema.validate_naming(cur)
        ]
        self.hierarchy.insert(0, ConfigSet.ROOT)
